woocommerce/woocommerce
# Fix getOption() not catch error properly

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:





We didn't catch the fetch error and rejected the promise properly in batch fetch control. That caused it to throw an unhandled promise rejection error, which could have crashed the core profiler.

This PR fixes the error handling and rewrites the code using async and await to improve readability.

Before:

<img width="1351" alt="Screenshot 2024-06-07 at 12 48 23" src="https://github.com/woocommerce/woocommerce/assets/4344253/7565f043-55f5-4469-9f25-a67dd73f40c4">

After:


<img width="1352" alt="Screenshot 2024-06-07 at 13 23 12" src="https://github.com/woocommerce/woocommerce/assets/4344253/3cd8b112-f52a-41bb-bd04-6271cc3ed5a4">


<img width="1351" alt="Screenshot 2024-06-07 at 12 48 53" src="https://github.com/woocommerce/woocommerce/assets/4344253/94a58608-789a-4f34-95e8-f676657a8fea">





### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Create a fresh site
2. Go to Core Profiler
3. Open dev tools > Network tab
4. Choose 'Fetch/XHR' filter and find the `options=woocommerce_allow_tracking` request or all option APIs
5. Right-click on the request and choose 'Block request URL'
6. Refresh the page
7. Confirm the page is loaded without any errors
8. Optionally, if you installed Redux DevTools, check the option store dispatches the SET_REQUESTING_ERROR action

<img width="1350" alt="Screenshot 2024-06-07 at 13 19 34" src="https://github.com/woocommerce/woocommerce/assets/4344253/3bed6768-fe51-4298-9c12-a5ab9bfad4f0">




### Changelog entry







<details>

<summary>Changelog Entry Details</summary>

#### Significance




#### Type




#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>
